src/etc/inc/util.inc

```diff
@@ -1932,24 +1932,63 @@ function mac_format($clientmac) {
 	}
 }
 
-function resolve_retry($hostname, $retries = 5) {
-
-	if (is_ipaddr($hostname)) {
-		return $hostname;
-	}
+function resolve_retry($hostname, $protocol = 'inet', $retries = 5, $numrecords = 1) {
 
+	$recresult = array();
+	$returnres = array();
 	for ($i = 0; $i < $retries; $i++) {
-		// FIXME: gethostbyname does not work for AAAA hostnames, boo, hiss
-		$ip = gethostbyname($hostname);
-
-		if ($ip && $ip != $hostname) {
-			/* success */
-			return $ip;
+		switch ($protocol) {
+			case 'any':
+				$checkproto = 'is_ipaddr';
+				$dnsproto = DNS_ANY;
+				$dnstype = array('A', 'AAAA');
+				break;
+			case 'inet6':
+				$checkproto = 'is_ipaddrv6';
+				$dnsproto = DNS_AAAA;
+				$dnstype = array('AAAA');
+				break;
+			case 'inet': 
+			default:
+				$checkproto = 'is_ipaddrv4';
+				$dnsproto = DNS_A;
+				$dnstype = array('A');
+				break;
+		}
+		if ($checkproto($hostname)) {
+			return $hostname;
+		}
+		$dnsresult = dns_get_record($hostname, $dnsproto);
+		foreach ($dnsresult as $dnsrec => $ip) {
+			if (is_array($ip)) {
+				if (in_array($ip['type'], $dnstype)) {
+				    if ($checkproto($ip['ip'])) { 
+					    $recresult[] .= $ip['ip'];
+				    }
+				    if ($checkproto($ip['ipv6'])) { 
+					    $recresult[] .= $ip['ipv6'];
+				    }
+				}
+			}
 		}
 
 		sleep(1);
 	}
 
+	if (!empty($recresult)) {
+		if ($numrecords == 1) {
+			return $recresult[0];
+		} else {
+			foreach (range(0, $numrecords-1) as $i) {
+				$rec = $recresult[$i];
+				if ($rec) {
+					$returnres[] .= $rec;
+				}
+			}
+			return $returnres;
```

Comment:
This whole block can be done in one line with a built-in function:

`return array_slice($recresult, 0, $numrecords);`

It's safe for `$numrecords` to be larger than the size of `$recresult` when using this function, it will return as many as it can.

Reply:
ok, done